### PR TITLE
Knockback API Edits

### PR DESCRIPTION
**Pull request by jkorn2324**
## Introduction

This pull request adds more flexibility to developers trying to add completely customizable knockback. Instead of a base knockback value that controls both the horizontal knockback (how far the entity goes after each hit) and the vertical knockback (how high the entity goes after each hit), this pull request separates the base horizontal and base vertical knockback values to allow for more customization without needing to create a Player class as shown in my plugin: https://github.com/jkorn2324/PvPCore. 

### Relevant issues

No issues so far.

### API changes

**Added Methods:**
EntityDamageByEntityEvent->getHorizontalKnockback();
EntityDamageByEntityEvent->getVerticalKnockback();
EntityDamageByEntityEvent->setHorizontalKnockback();
EntityDamageByEntityEvent->setVerticalKnockback();

**Changed Methods:**
EntityDamageByEntityEvent->setKnockBack();
EntityDamageByEntityEvent->getKnockBack();
Living->knockBack();

### Behavioural changes

No Behavioural changes, just API changes.

## Backwards compatibility

This is backwards-compatible with the current API.

## Follow-up

Determine whether the function names in EntityDamageByEntityEvent are clear enough and whether or not I even need the setVerticalKnockback/setHorizontalKnockback functions if the function names are confusing.

## Tests

Tested by changing the Horizontal Knockback, Vertical Knockback in a plugin using this updated API and by going on a test server and having two players hit each other.

API Tests Video Link:
https://www.dropbox.com/s/ls0kqq04bke17jk/API%20Tests.MOV?dl=0

### CODE REVIEW

**Review by dktapps — CHANGES_REQUESTED** [review]
The formatting in this PR is a mess.

This is also an alarmingly large amount of changes for what's essentially just allowing to customise the vertical knockback limit.

**Comment by jkorn2324** [comment]
What do you mean by formatting in this context? Do you have any specific critiques to make it more organized?

**Comment by dktapps** [comment]
image
image
image

There's a bunch of other minor spacing issues, but the indentation here is awful.

File: src/pocketmine/entity/Living.php [hunk 1]
```diff
@@ -576,25 +580,28 @@ protected function doHitAnimation() : void{
 		$this->broadcastEntityEvent(ActorEventPacket::HURT_ANIMATION);
 	}
 
-	public function knockBack(Entity $attacker, float $damage, float $x, float $z, float $base = 0.4) : void{
-		$f = sqrt($x * $x + $z * $z);
+	public function knockBack(Entity $attacker, float $damage, float $x, float $z, float $base = 0.4, ?float $baseY = null) : void{
+	    $f = sqrt($x * $x + $z * $z);
```
Comment (jasonwynn10) [inline src/pocketmine/entity/Living.php:584]: tabs vs spaces

File: src/pocketmine/entity/Living.php [hunk 2]
```diff
@@ -576,25 +580,28 @@ protected function doHitAnimation() : void{
...
 		if($f <= 0){
 			return;
 		}
+
```
Comment (jasonwynn10) [inline src/pocketmine/entity/Living.php:588]: new lines inserted

File: src/pocketmine/event/entity/EntityDamageByEntityEvent.php [hunk 3]
```diff
@@ -65,11 +69,67 @@ public function getDamager() : ?Entity{
...
-	public function setKnockBack(float $knockBack) : void{
-		$this->knockBack = $knockBack;
+	/**
+	 * Sets the horizontal knockback of the event directly. Added for clarification purposes.
+	 * @param float $horizontalKnockBack - The horizontal knockback of the event.
+	 */
+	public function setHorizontalKnockBack(float $horizontalKnockBack): void{
+	    // Sets the vertical knockback to the previous horizontal knockback.
+	    if($this->verticalKnockBack === null){
+            $this->verticalKnockBack = $this->horizontalKnockBack;
```
Comment (jasonwynn10) [inline src/pocketmine/event/entity/EntityDamageByEntityEvent.php:122]: tabs vs spaces

File: src/pocketmine/event/entity/EntityDamageByEntityEvent.php [hunk 4]
```diff
@@ -31,17 +31,21 @@
  * Called when an entity takes damage from another entity.
  */
 class EntityDamageByEntityEvent extends EntityDamageEvent{
+
```
Comment (jasonwynn10) [inline src/pocketmine/event/entity/EntityDamageByEntityEvent.php:34]: random new line

File: src/pocketmine/entity/Living.php [hunk 5]
```diff
@@ -554,15 +557,16 @@ public function attack(EntityDamageEvent $source) : void{
 		if($source instanceof EntityDamageByChildEntityEvent){
 			$e = $source->getChild();
 			if($e !== null){
-				$motion = $e->getMotion();
-				$this->knockBack($e, $source->getBaseDamage(), $motion->x, $motion->z, $source->getKnockBack());
+                $motion = $e->getMotion();
+                $this->knockBack($e, $source->getBaseDamage(), $motion->x, $motion->z, $source->getHorizontalKnockBack(), $source->getVerticalKnockBack());
 			}
+
```
Comment (jasonwynn10) [inline src/pocketmine/entity/Living.php:563]: random new line

**Review by jasonw4331 — CHANGES_REQUESTED** [review]
Please fix the formatting

File: src/pocketmine/entity/Living.php (hunk 5 above)
Comment (jkorn2324) [reply src/pocketmine/entity/Living.php]: > Please fix the formatting

Added your suggestions

File: src/pocketmine/entity/Living.php [hunk 6]
```diff
@@ -533,14 +533,17 @@ public function attack(EntityDamageEvent $source) : void{
 
 		$this->applyDamageModifiers($source);
 
-		if($source instanceof EntityDamageByEntityEvent and (
-			$source->getCause() === EntityDamageEvent::CAUSE_BLOCK_EXPLOSION or
+		if ($source instanceof EntityDamageByEntityEvent and
```
[the remark this answers is not in the document]
Comment (jasonwynn10) [reply src/pocketmine/entity/Living.php]: also some tabs were changed to spaces

**Comment by CortexPE** [comment]
This could be done by plugins, like I've done myself without changing `Player` class more than two years ago.
Gotta admit that it is very hacky but it's possible without all these extra code. Doesn't even use reflection...
Not sure if I could disclose the details on how I did it even without modifying PM itself but it is indeed possible.

In my method, you'd just really have to reimplement the knockback logic yourself... which on the plus side, does give you full customizability on it

**Comment by dktapps** [comment]
This is stale, overcomplicated, and breaks backwards compatibility.

In a nutshell, what's needed here is a way to limit the maximum vertical knockback. Currently, vertical knockback limit scales with the knockback force, which is not supposed to be the case (as in vanilla). Implementing this should be much simpler than the changes seen in this PR, minus the fact that adding an extra parameter to Living->knockBack() would also break BC.

Considering the age of this problem and the ramifications for BC, I think a solution is better targeted at PM4.

**Pull request closed without merging**